### PR TITLE
avoid back trace when importing preferred communication method

### PR DESCRIPTION
**Pull request by jmcclelland**
Overview
----------------------------------------

When importing a contact with preferred communication method set, I get:

CiviCRM_API3_Exception: "'' is not a valid option for field preferred_communication_method"

The problem surfaced after upgrading from 5.33 to 5.39.

I traced it back to this code and can fix it with this simple change.

The problem with the original code is that `$formatted[$fieldName]` is an
array with the key set to the chosen value and the value set to 1. So, the array_search always fails since the key it's looking for is an array, not a string.

I think the line:

`&& empty($fieldSpec['options'][$formatted[$fieldName]])) {`

Is also not quite right but not sure how to change it.

Before
----------------------------------------

Importing a contact with preferred communication method set results in:

'' is not a valid option for field preferred_communication_method

After
----------------------------------------

The contact record is imported.

Technical Details
----------------------------------------
The problem may have been introduced via cdfa6649f87d29341cc680e5d42991b23283a238.

Comments
----------------------------------------

I don't think this fix is quite there yet but would love to get some feedback.

Also, in this particular use case, the label for the SMS option has been changed to "Text message" but I don't think that has an impact on this issue.

### CODE REVIEW

**Comment by jmcclelland** [comment]
Woops - one more commit + squash (I accidentally left in my incorrect earlier fix).

File: tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php [hunk 1]
```diff
@@ -850,6 +850,45 @@ public function testImportAmbiguousStateCountry(): void {
...
+    $processor->setContactType('Individual');
+    $processor->setMappingFields(
+      [
+        ['name' => 'first_name'],
+        ['name' => 'last_name'],
+        ['name' => 'preferred_communication_method'],
+        ['name' => 'gender_id'],
+        ['name' => 'preferred_language'],
+      ],
+    );
```
Comment (demeritcowboy) [inline tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php:872]: Console shows `PHP Parse error:  syntax error, unexpected ')' in tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php on line 872`.

Comment (jmcclelland) [reply tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php:872]: Ah - Thank you! I think I added that in my last minute `civilint` efforts and didn't notice that I added a comma where it doesn't belong.

Comment (jmcclelland) [reply tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php:872]: @demeritcowboy thanks for the feedback - now this seems to be passing with that syntax fix.

File: tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php [hunk 2]
```diff
@@ -850,6 +850,45 @@ public function testImportAmbiguousStateCountry(): void {
...
+    $fields = ['Ima', 'Texter', "SMS,Phone", "Female", "Danish"];
+    $importer->import(CRM_Import_Parser::DUPLICATE_NOCHECK, $fields);
+    $contact = $this->callAPISuccessGetSingle('Contact', ['last_name' => 'Texter']);
+
+    $this->assertEquals([4, 1], $contact['preferred_communication_method'], "Import multiple preferred communication methods using labels.");
+    $this->assertEquals(1, $contact['gender_id'], "Import gender with label.");
+    $this->assertEquals('da_DK', $contact['preferred_language'], "Import preferred language with label.");
+
+    $importer = $processor->getImporterObject();
+    $fields = ['Ima', 'Texter', "4,1", "1", "da_DK"];
```
Comment (demeritcowboy) [inline tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php:883]: It's interesting that this works in the test, because I can't get numbers to work for this field in the UI, either before or after the patch or using single-value or multi.
Mystery?

**Comment by demeritcowboy** [comment]
* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
      * Tried a couple different things including custom multivalue fields. There's just the comment about values for this communication field, but it didn't work before the patch either.
* Developer standards
    * [`r-tech`] __PASS__
      * The code is hard to follow already, but the comment maybe helps.
    * [`r-code`] __PASS__
    * [`r-maint`] __PASS__
    * [`r-test`] __PASS__
      * That the test works but UI doesn't might just be that the prevalidation is wrong but the import itself is ok.

File: tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php (hunk 2 above)
Comment (jmcclelland) [reply tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php:883]: Huh - interesting. I put the values into the test because I thought, why not expand the test coverage? But didn't test entering the values in the UI. What happens when you try? Do you get an error?

Comment (demeritcowboy) [reply tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php:883]: You can't get past the import screens to see what would happen - it gives you the excel file to download with errors which says it thinks the field has an invalid value.

**Comment by jmcclelland** [comment]
@demeritcowboy - ahhh.... see that now. The test code is executing the same code as the UI, but the code doesn't consider it a fatal error (it's more of an FYI) so the test code happily continues. I think it's good for the test to remain and a new issue could be opened to:

1. Fix all import testing code to check for this kind of error
2. Fix `CRM_Contact_Import_Parser_Contact::isErrorInCoreData` to allow values for this field (and it looks like a few others).
